internal/workshare/workshare_waitgroup.go

```diff
@@ -52,19 +53,27 @@ import (
 type AsyncGroup struct {
 	wg       *sync.WaitGroup
 	requests []interface{}
+	waited   bool
 }
 
 // Wait waits for scheduled asynchronous work to complete and returns all asynchronously processed inputs.
 func (g *AsyncGroup) Wait() []interface{} {
-	if g.wg == nil {
+	if g.wg == nil || g.waited {
 		return nil
 	}
 
 	g.wg.Wait()
 
+	g.waited = true
+
 	return g.requests
 }
 
+// Close ensures all asynchronous work has been awaited for.
+func (g *AsyncGroup) Close() {
+	g.Wait()
+}
+
 // RunAsync starts the asynchronous work to process the provided request,
 // the user must call Wait() after all RunAsync() have been scheduled.
 func (g *AsyncGroup) RunAsync(w *Pool, process ProcessFunc, request interface{}) {
```

Comment:
are `AsyncGroup`s supposed to be able to `Wait()` more than once? It seems the below if-block (I can't comment directly on it) should either be
```
if g.waited {
  panic("invalid usage: AsyncGroup cannot be reused after Wait()")
}

if g.wg == nil {
  ...
```
or 
```
if g.wg == nil {
  waited = false
 ...
```

Otherwise, I think it'll skip waiting properly when the group is reused

Reply:
added flags to disallow double `Wait()` and `Wait()` afet `Close()`. Explicitly allowed redundant `Close()` after `Close()`